preprocess/foldfilter_main.cc

```diff
@@ -0,0 +1,320 @@
+	int32_t length = strlen(value);
+	int32_t pos = 0;
+	
+	vector<UChar32> delimiters;
+	delimiters.reserve(length);
+
+	while (pos < length) {
+		UChar32 delimiter;
+		U8_NEXT(value, pos, length, delimiter);
+		
+		if (delimiter < 0)
+			throw utf8::NotUTF8Exception(value);
+		
+		delimiters.push_back(delimiter);
+	}
+
+	return delimiters;
+}
+
+void parse_options(program_options &options, int argc, char **argv) {
+	while (true) {
+		switch(getopt(argc, argv, "+w:d:sh")) {
+			case 'w':
+				options.column_width = atoi(optarg);
+				continue;
+
+			case 'd':
+				options.delimiters = parse_delimiters(optarg);
+				continue;
+
+			case 's':
+				options.keep_delimiters = false;
+				continue;
+
+			case 'h':
+			case '?':
+			default:
+				exit(usage(argv));
+
+			case -1:
+				break;
+		}
+		break;
+	}
+
+	if (optind == argc)
+		exit(usage(argv));
+
+	options.child_argv = argv + optind;
+}
+
+int main(int argc, char **argv) {
+	program_options options;
+
+	parse_options(options, argc, argv);
+
+	util::UnboundedSingleQueue<deque<string>> queue;
+
+	util::scoped_fd child_in_fd, child_out_fd;
+
+	pid_t child = preprocess::Launch(options.child_argv, child_in_fd, child_out_fd);
+
+	thread feeder([&child_in_fd, &queue, &options]() {
+		util::FilePiece in(STDIN_FILENO);
+		util::FileStream child_in(child_in_fd.get());
+
+		for (StringPiece sentence : in) {
+			deque<StringPiece> lines;
+			deque<string> delimiters;
+
+			// If there is nothing to wrap, it will end up with a single line
+			// and a single empty delimiter.
+			tie(lines, delimiters) = wrap_lines(sentence, options);
+			// assert(lines.size() == delimiters.size());
+
+			// When we're keeping delimiters all of these will be empty strings
+			// but their amount at least will tell the reader thread how many
+			// lines it needs to consume to reconstruct the single line.
+			queue.Produce(delimiters);
+
+			// Feed the document to the child.
+			// Might block because it can cause a flush.
+			for (auto const &line : lines)
+				child_in << line << '\n';
+		}
+
+		// Tell the reader to stop
+		queue.Produce(deque<string>());
+
+		// Flush (blocks) & close the child's stdin
+		child_in.flush();
+		child_in_fd.reset();
+	});
+
+	thread reader([&child_out_fd, &queue, &options]() {
+		util::FileStream out(STDOUT_FILENO);
+		util::FilePiece child_out(child_out_fd.release());
+
+		deque<string> delimiters;
+		string sentence;
+
+		for (size_t sentence_num = 1; queue.Consume(delimiters).size() > 0; ++sentence_num) {
+			sentence.clear();
+			
+			// Let's assume that the wrapped process plus the chopped off
+			// delimiters won't be more than twice the input we give it.
+			sentence.reserve(delimiters.size() * 2 * options.column_width);
+
+			try {
+				while (!delimiters.empty()) {
+					StringPiece line(child_out.ReadLine());
+					sentence.append(line.data(), line.length());
+					sentence.append(delimiters.front());
+					delimiters.pop_front();
```

Comment:
I think we're using `deque` for this `pop_front` here but you're not saving much at all compared to iterating a vector.

Reply:
I think my reasoning was that I don't need random access and I don't know how many items will be going into it, so lets go for deque because then at least the push_back will be cheap. But push_back on a vector isn't all that expensive either… and now it has to do a lot more pointer dereferencing. Hmm…